src/base/misc/disks.c

```diff
@@ -929,163 +916,100 @@ static void image_setup(struct disk *dp)
   }
 }
 
-/* XXX - relies upon a file of SECTOR_SIZE in PARTITION_PATH that which
- *       contains the MBR (first sector) of the drive (i.e. /dev/hda).
- *       only works for /dev/hda1 right now, and only allows one
- *       partition, which must begin at head 1, sec 1, cyl 0 (i.e. linear
- *       sector 2, the second one).
- *       Also, it eats memory proportional to the number of sectors before
- *       the start of the partition.
- */
-
 static void partition_auto(struct disk *dp)
 {
+  unsigned char sect[0x200];
+  struct on_disk_vbr *vbr = (struct on_disk_vbr *)sect;
+
   d_printf("PARTITION auto\n");
-}
 
-static void partition_setup(struct disk *dp)
-{
+  if (!dp->part_image) {
 #ifdef __linux__
-  int part_fd, i;
-  unsigned char tmp_mbr[SECTOR_SIZE];
-  char *hd_name;
-  struct on_disk_partition p;
-
-#define PNUM dp->part_info.number
-
-  /* PNUM is 1-based */
-
-  d_printf("PARTITION SETUP for %s\n", dp->dev_name);
-
-  hd_name = strdup(dp->dev_name);
-  hd_name[8] = '\0';			/* i.e.  /dev/hda6 -> /dev/hda */
-
-  part_fd = SILENT_DOS_SYSCALL(open(hd_name, O_RDONLY | O_CLOEXEC));
-  if (part_fd == -1) {
-    if (dp->floppy) {
-      free(hd_name);
-      return;
+    unsigned long length;
+    if (ioctl(dp->fdesc, BLKGETSIZE, &length)) {
+      error("calling ioctl BLKGETSIZE for PARTITION %s\n", dp->dev_name);
+      leavedos(22);
     }
-    PNUM = 1;
-    set_part_ent(dp, tmp_mbr);
-    tmp_mbr[0x1fe] = 0x55;
-    tmp_mbr[0x1ff] = 0xaa;
+    dp->num_secs = length;
+#else
+    error("no support for block device PARTITION on non linux platforms\n");
+    leavedos(22);
+#endif
   } else {
-    (void)RPT_SYSCALL(read(part_fd, tmp_mbr, SECTOR_SIZE));
-    close(part_fd);
-    d_printf("Using MBR from %s for PARTITION %s (part#=%d).\n",
-             hd_name, dp->dev_name, PNUM);
+    struct stat sb;
+    if (fstat(dp->fdesc, &sb)) {
+      error("calling ioctl FSTAT for PARTITION(image) %s\n", dp->dev_name);
+      leavedos(22);
+    }
+    dp->num_secs = sb.st_size / SECTOR_SIZE;
   }
-  free(hd_name);
 
-  /* check for logical partition, if so simulate as primary part#1 */
-  if (PNUM > 4 ) {
-    d_printf("LOGICAL PARTITION - will be simulated as physical partition 1\n");
-    PNUM = 1;
-    set_part_ent(dp, tmp_mbr);
+  lseek(dp->fdesc, 0, SEEK_SET);
+  if (RPT_SYSCALL(read(dp->fdesc, sect, sizeof(sect))) != sizeof(sect)) {
```

Comment:
Well, that will indeed make the
code cleaner. Here you read to
some `sect` and immediately
access `vbr`, which makes the
aliasing violation visible even
during a review.